File: 0222-count-complete-tree-nodes/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int getDepth(TreeNode* root)
    {
        int depth = 0;
        while(root->left != nullptr)
        {
            depth++;
            root = root->left;
        }
        
        return depth;
    }
    
    bool exists(int depth, TreeNode* root, int index)
    {
        int left  = 0;
        int right = pow(2, depth) - 1;
        
        for(int iter = 0; iter < depth; iter++)
        {
            int mid = left + (right - left)/2;
            
            if(index <= mid)
            {
                root = root->left;
                right = mid;
            }
            else
            {
                root = root->right;
                left = mid + 1;
            }            
        }
    
        return (root != nullptr);
    }
    
public:
    int countNodes(TreeNode* root) 
    {
        if(root == nullptr) return 0;
        
        int depth = getDepth(root);
        if(depth == 0) return 1;
        
        int left  = 1;
        int right = pow(2, depth) - 1;
        
        
        while(left <= right)
        {
            int mid = left + (right - left)/2;
            
            if(exists(depth, root, mid))
                left = mid + 1;
            else
                right = mid - 1;
        }
        
        return pow(2, depth) - 1 + left;
        
        
    }
};

File: 0222-count-complete-tree-nodes/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getDepth(self, node: Optional[TreeNode]) -> int:
        d = 0
        while node.left:
            node = node.left
            d += 1
        return d
    
    def exists(self, index:int, d: int, node: TreeNode)->bool:
        left, right = 0, 2**d - 1
        
        for _ in range(d):
            mid = left + (right - left)//2
            
            if index <= mid:
                node  = node.left
                right = mid
            else:
                node = node.right
                left = mid + 1
        
        return True if node else False
            
    def countNodes(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        d = self.getDepth(root)
        
        if d == 0: 
            return 1
        
        left, right = 1, 2**d - 1
        
        while left <= right:
            mid = left + (right - left)//2
            
            if self.exists(mid, d, root):
                left = mid + 1
            else:
                right = mid - 1
        
        return (2**d - 1) + left
    
